contracts/registries/PoolRegistry.sol

```diff
@@ -0,0 +1,438 @@
+    /// @inheritdoc IPoolRegistry
+    function getPaused(address poolAddress)
+        external
+        view
+        override
+        hasMatchingPool(poolAddress)
+        returns (bool)
+    {
+        return ISwap(poolAddress).paused();
+    }
+
+    /// @inheritdoc IPoolRegistry
+    function getSwapStorage(address poolAddress)
+        external
+        view
+        override
+        hasMatchingPool(poolAddress)
+        returns (SwapStorageData memory swapStorageData)
+    {
+        swapStorageData = pools[poolsIndexOfPlusOne[poolAddress] - 1].isGuarded
+            ? _getSwapStorageGuarded(poolAddress)
+            : _getSwapStorage(poolAddress);
+    }
+
+    function _getSwapStorage(address poolAddress)
+        internal
+        view
+        returns (SwapStorageData memory swapStorageData)
+    {
+        (
+            swapStorageData.initialA,
+            swapStorageData.futureA,
+            swapStorageData.initialATime,
+            swapStorageData.futureATime,
+            swapStorageData.swapFee,
+            swapStorageData.adminFee,
+            swapStorageData.lpToken
+        ) = ISwap(poolAddress).swapStorage();
+    }
+
+    function _getSwapStorageGuarded(address poolAddress)
+        internal
+        view
+        returns (SwapStorageData memory swapStorageData)
+    {
+        (
+            swapStorageData.initialA,
+            swapStorageData.futureA,
+            swapStorageData.initialATime,
+            swapStorageData.futureATime,
+            swapStorageData.swapFee,
+            swapStorageData.adminFee,
+            ,
+            swapStorageData.lpToken
+        ) = ISwapGuarded(poolAddress).swapStorage();
+    }
+
+    /// @inheritdoc IPoolRegistry
+    function getTokens(address poolAddress)
+        external
+        view
+        override
+        hasMatchingPool(poolAddress)
+        returns (address[] memory)
+    {
+        uint256 poolIndex = poolsIndexOfPlusOne[poolAddress];
+        return pools[poolIndex - 1].tokens;
+    }
+
+    /// @inheritdoc IPoolRegistry
+    function getUnderlyingTokens(address poolAddress)
+        external
+        view
+        override
+        hasMatchingPool(poolAddress)
+        returns (address[] memory)
+    {
+        uint256 poolIndex = poolsIndexOfPlusOne[poolAddress];
+        return pools[poolIndex - 1].underlyingTokens;
+    }
+
+    /// @inheritdoc IPoolRegistry
+    function getPoolsLength() external view override returns (uint256) {
+        return pools.length;
+    }
+
+    /// @inheritdoc IPoolRegistry
+    function getEligiblePools(address from, address to)
+        external
+        view
+        override
+        returns (address[] memory eligiblePools)
+    {
+        require(
+            from != address(0) && from != to,
+            "PR: from and to cannot be the zero address"
+        );
+        return eligiblePairsMap[uint160(from) ^ uint160(to)];
+    }
+
+    /// @inheritdoc IPoolRegistry
+    function getBalances(address poolAddress)
```

Comment:
I think we also need a `getUnderlyingBalances`. Also what are your thoughts of `getBalances` returning purely balances since we already have a `getTokens` function? I think we could do a `getTokensAndBalances` fn if we wanted both

Reply:
Sounds good I can split these up to be more explicit